server/queues/tasks/DocumentAddUserNotificationsTask.ts

```diff
@@ -1,27 +1,62 @@
-import { NotificationEventType } from "@shared/types";
+import { DocumentPermission, NotificationEventType } from "@shared/types";
+import Logger from "@server/logging/Logger";
 import { Notification, User } from "@server/models";
 import { DocumentUserEvent } from "@server/types";
 import BaseTask, { TaskPriority } from "./BaseTask";
 
 export default class DocumentAddUserNotificationsTask extends BaseTask<DocumentUserEvent> {
   public async perform(event: DocumentUserEvent) {
-    const recipient = await User.findByPk(event.userId);
-    if (!recipient) {
+    const permission = event.changes?.attributes.permission as
+      | DocumentPermission
+      | undefined;
+
+    if (!permission) {
+      Logger.debug(
+        "task",
+        `Suppressing notification for user ${event.userId} as permission not available`,
+        {
+          documentId: event.documentId,
+          userId: event.userId,
+        }
+      );
       return;
     }
 
+    const recipient = await User.findByPk(event.userId);
     if (
-      !recipient.isSuspended &&
-      recipient.subscribedToEventType(NotificationEventType.AddUserToDocument)
+      !recipient ||
+      recipient.isSuspended ||
+      !recipient.subscribedToEventType(NotificationEventType.AddUserToDocument)
     ) {
-      await Notification.create({
-        event: NotificationEventType.AddUserToDocument,
-        userId: event.userId,
-        actorId: event.actorId,
-        teamId: event.teamId,
-        documentId: event.documentId,
-      });
+      return;
     }
+
+    const hasHigherPermission = await recipient.hasHigherDocumentPermission({
+      documentId: event.documentId,
+      permission,
+      skipMembershipId: event.modelId,
+    });
+
+    if (hasHigherPermission) {
```

Comment:
So this suppresses if they already have a higher permission, what if they have the same permission already – shouldn't we suppress in that situation too? 🤔